core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala

```diff
@@ -43,26 +44,16 @@ private[spark] trait ShuffleManager {
       context: TaskContext,
       metrics: ShuffleWriteMetricsReporter): ShuffleWriter[K, V]
 
-  /**
-   * Get a reader for a range of reduce partitions (startPartition to endPartition-1, inclusive).
-   * Called on executors by reduce tasks.
-   */
-  def getReader[K, C](
-      handle: ShuffleHandle,
-      startPartition: Int,
-      endPartition: Int,
-      context: TaskContext,
-      metrics: ShuffleReadMetricsReporter): ShuffleReader[K, C]
-
   /**
    * Get a reader for a range of reduce partitions (startPartition to endPartition-1, inclusive) to
-   * read from map output (startMapIndex to endMapIndex - 1, inclusive).
+   * read from a range of map outputs which specified by mapIndexRange(startMapIndex to
+   * endMapIndex-1, inclusive).
+   *
    * Called on executors by reduce tasks.
    */
-  def getReaderForRange[K, C](
+  def getReader[K, C](
       handle: ShuffleHandle,
-      startMapIndex: Int,
-      endMapIndex: Int,
+      mapIndexRange: Array[MapStatus] => (Int, Int),
```

Comment:
It's a function only because we need to know the total number of mappers? Maybe we can still use `startMapIndex: Int` and `endMapIndex: Int`, caller side can use `Int.Max` as `endMapIndex`, to indicate that it wants to read all the mappers.

Reply:
> It's a function only because we need to know the total number of mappers?

Yes.

Just a little worry that `Int.Max` is not easy to know from intuitive. It also bring small trouble for log info in `getMapSizesByExecutorId` since you can't know the real `endMapIndex` there.

Reply:
We can document it, it's an internal API anyway.

For logging, we do know the number of total mappers inside the method, right?

Reply:
Ok, I get it how to know the `endMapIndex`.